third_party/VulkanTutorial/OffscreenRenderingVulkanTutorial.cpp

```diff
@@ -0,0 +1,575 @@
+  CreateImageView();
+  CreateRenderPass();
+  CreateGraphicsPipeline();
+  CreateFramebuffer();
+  CreateCommandPool();
+  CreateCommandBuffer();
+  CreateFence();
+}
+
+void OffscreenRenderingVulkanTutorial::MainLoop(uint64_t frame_count) {
+  LOG("MainLoop");
+  for (uint64_t frame = 0; frame < frame_count; ++frame) {
+    absl::Time next_frame_time = absl::Now() + absl::Microseconds(16667);
+    DrawFrame();
+
+    {
+      absl::MutexLock stop_requested_lock{&stop_requested_mutex_};
+      if (stop_requested_mutex_.AwaitWithDeadline(absl::Condition(&stop_requested_),
+                                                  next_frame_time)) {
+        break;
+      }
+    }
+  }
+  vkDeviceWaitIdle(device_);
+}
+
+void OffscreenRenderingVulkanTutorial::CleanUp() {
+  LOG("CleanUp");
+  vkDestroyFence(device_, fence_, nullptr);
+  // Command buffers will be automatically freed when their command pool is destroyed, so we don't
+  // need an explicit cleanup.
+  vkDestroyCommandPool(device_, command_pool_, nullptr);
+  vkDestroyFramebuffer(device_, framebuffer_, nullptr);
+  vkDestroyPipeline(device_, graphics_pipeline_, nullptr);
+  vkDestroyPipelineLayout(device_, pipeline_layout_, nullptr);
+  vkDestroyRenderPass(device_, render_pass_, nullptr);
+  vkDestroyImageView(device_, image_view_, nullptr);
+  vkFreeMemory(device_, memory_, nullptr);
+  vkDestroyImage(device_, image_, nullptr);
+  vkDestroyDevice(device_, nullptr);
+  vkDestroyInstance(instance_, nullptr);
+}
+
+void OffscreenRenderingVulkanTutorial::CreateInstance() {
+  VkApplicationInfo app_info{
+      .sType = VK_STRUCTURE_TYPE_APPLICATION_INFO,
+      .pApplicationName = "VulkanTutorial",
+      .applicationVersion = VK_MAKE_VERSION(1, 0, 0),
+      .pEngineName = "",
+      .engineVersion = VK_MAKE_VERSION(1, 0, 0),
+      .apiVersion = VK_API_VERSION_1_0,
+  };
+
+  CHECK(AreValidationLayersSupported());
+  VkInstanceCreateInfo create_info{
+      .sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO,
+      .pApplicationInfo = &app_info,
+      .enabledLayerCount = static_cast<uint32_t>(kValidationLayerNames.size()),
+      .ppEnabledLayerNames = kValidationLayerNames.data(),
+      // Note that, unlike what is described at
+      // https://vulkan-tutorial.com/Drawing_a_triangle/Setup/Instance, we don't need to enable
+      // extensions for the window system, as we are performing offscreen rendering.
+      .enabledExtensionCount = 0,
+      .ppEnabledExtensionNames = nullptr,
+  };
+
+  CHECK_VK_SUCCESS(vkCreateInstance(&create_info, nullptr, &instance_));
+}
+
+bool OffscreenRenderingVulkanTutorial::AreValidationLayersSupported() {
+  uint32_t layer_count = 0;
+  CHECK_VK_SUCCESS(vkEnumerateInstanceLayerProperties(&layer_count, nullptr));
+
+  std::vector<VkLayerProperties> available_layers(layer_count);
+  CHECK_VK_SUCCESS(vkEnumerateInstanceLayerProperties(&layer_count, available_layers.data()));
+
+  for (const char* layer_name : kValidationLayerNames) {
```

Comment:
Theoretically, you might also want to check for the version.

Reply:
Do you know what I should check exactly? That `layer_properties.specVersion >= VK_API_VERSION_1_1`?

Reply:
For your concrete layer there is the `VK_EXT_VALIDATION_FLAGS_SPEC_VERSION`. But I am not 100% sure if this is really suggested to check for layers. 

For extensions there is something similar, and there it is suggested.

Reply:
As discussed offline, let's leave as it is.